DaOnPark

deepseafishy opened the pull request:
Added singleton pattern to AccountCustomRepository & MetadataCustomRepository

deepseafishy committed:
Repository들에 Singleton 디자인 패턴 적용

많은 new 선언을 없애기 위한 singleton 디자인 패턴 적용
@TODO: MetadataCustomRepository에 collection 관련 수정 필요.

AccountCustomRepository.java
1. instance라는 변수를 만들어 singleton 디자인 패턴 적용.
2. getCurrentAccount() 함수 구현.

MetadataCustomRepository.java
1. instance라는 변수를 만들어 singleton 디자인 패턴 적용.

deepseafishy committed:
Merge branch 'master' of https://github.com/pnu-004-team4/team04 into DaOnPark

deepseafishy committed:
Create JUnit test cases for all db classes

Creation of JUnit test cases for all db classes for increasing code coverage.

deepseafishy committed:
Adding singleton pattern to DB repository class: Hashmap implementation

Process of adding singleton pattern to DB repository. Since MetadataCustomRepository will need instances of each user, a hash map has been implemented to do so. This step would be the last step if everything works out.

MetadataCustomRepository.java
1. Hashmap has been added as a private variable.
2. Changes to getInstance() method to look and find, or create a new instance of MetadataCustomRepository.

deepseafishy committed:
Merge branches 'DaOnPark' and 'master' of https://github.com/pnu-004-team4/team04 into DaOnPark

deepseafishy committed:
Added singleton pattern to repository classes

Applied singleton pattern to AccountCustomRepository and MetadataCustomRepository. This step involves fetching latest files, and checking if other classes have errors due to changes made in said files.

AccounCustomRepository.java
1. Applied a singleton pattern.
2. Due to singleton pattern being applied, getInstance() method was created.

MetadataCustomRepository.java
1. Applied a singleton pattern.
2. Due to singleton pattern being applied, getInstance() method was created.
3. MetadataCustomRepository needs to be instantiated for each users, thus hash map was implemented instead of a single private variable.

deepseafishy committed:
Commented out JUnit test files for codacy and travisCLI

deepseafishy committed:
Commented out JUnit test files for codacy and travisCLI v2

Completely commented out test files.

deepseafishy committed:
Commented out JUnit test files for codacy and travisCLI v3

Added private to variables in converter test files.

deepseafishy committed:
Commented out JUnit test files for codacy and travisCLI v4

Deleted unused imports in AudioMetaConverterTest.

deepseafishy committed:
Commented out JUnit test files for codacy and travisCLI v5

Commented out testToAudioMetaException() due to Travis CI build error.

BlaCkinkGJ reviewed (approved):
test 구문이라서 codacy를 통과하지 못했지만 무시하고 진행하면 될 것 같습니다.

deepseafishy committed:
Merge branch 'master' into DaOnPark

# Conflicts:
#	src/main/java/com/team04/musiccloud/db/AccountCustomRepository.java

deepseafishy committed:
Creating JUnit test cases for code coverage

Implemented JUnit test cases for code coverage. Haven't covered all the parts, but all of them are covered at least 75%. Additional code refactoring were made to carry out test cases.

AccountConverter.java
1. Error was shown in parsing String to boolean, therefore little adjustment was made from casting to String to using toString() method.

AudioMetaConverter.java
1. Whenever appending dbId to a document, it needs to be a new ObjectID() or else it will be stored as a String. Thus little adjustment was made from appending audioMeta.getDbId() to new ObjectId(audioMeta.getDbId());.

AccountDao.java
1. getAccount() method changed a little to check if the search result is null or not. If it is null, it will return null instead of sending null value to AccountConverter.toAccount(document);

AudioMetaDao.java
1. Deleted deleteDataInDatabase() method.
2. Changed delete() method to use Filter.

FileMetaDao.java
1. Changed delete() method to use Filter.

AccountCustomRepository.java
1. getInstance() method is now a static method, thus instance variable also became static variable.
2. updateAccount() method will now check if search result for an account returns null or not. If it returns null, then updateAccount will return false.

MetadataCustomRepository.java
1. getInstance() method is now a static method, thus instance variable also became static variable.

BlaCkinkGJ reviewed (changes requested):
(중복 내용이므로 삭제)

BlaCkinkGJ commented on src/test/java/com/team04/musiccloud/db/converter/AudioMetaConverterTest.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Test
+    public void testToDocument() {
+        assertEquals(document.toString(), toDocument(audioMeta).toString());
```
이 부분에서 아래와 같이 Travis CI에서 문제가 발생했다고 알려주고 있습니다.
```java
com.team04.musiccloud.db.converter.AudioMetaConverterTest > testToDocument FAILED
    java.lang.IllegalArgumentException at AudioMetaConverterTest.java:51
```
이 말은 **_Document 생성 과정에서 잘못_** 되었다는 것을 의미합니다.

BlaCkinkGJ commented on src/test/java/com/team04/musiccloud/db/converter/AudioMetaConverterTest.java:
```diff
@@ -0,0 +1,75 @@
+                .append("album", "album")
+                .append("releaseDate", now.toString())
+                .append("durationMs", "0")
+                .append("count", "0")
```
저는 이 부분이 문제를 야기하는 부분으로 생각합니다.

이미 `AudioMetaConverter.java` 파일에서 정의하는 `toDocument`의 내용을 보면 아래와 같습니다.
```java
  public static Document toDocument(AudioMeta audioMeta) {
    Document document = new Document("title", audioMeta.getTitle())
        .append("author", audioMeta.getAuthor())
        .append("album", audioMeta.getAlbum())
        .append("releaseDate", audioMeta.getReleaseDate().toString())
        .append("durationMs", audioMeta.getDurationMs())
        .append("count", audioMeta.getPlayCount());
    if (audioMeta.getDbId() == null) {
      document.append("_id", new ObjectId());
    } else {
      document.append("_id", audioMeta.getDbId());
    }

    return document;
  }
```
여기서 괄목할 부분은 **`getDurationMs()`와 `getPlayCount()` 입니다. 해당 부분이 `int` 형을 반환**하는 것을 `AudioMeta.java` 파일을 확인을 하면 알 수 있습니다.

그런데, 이 부분에서는 `String`으로 `append`를 수행하고 있습니다.

BlaCkinkGJ reviewed (changes requested):
현재 발생하는 문제를 **아래와 같이 고치면 문제가 해결되지 않을까 생각합니다.**
```java
        this.document = new Document()
                .append("title", "title")
                .append("author", "author")
                .append("album", "album")
                .append("releaseDate", now.toString())
                .append("durationMs", 0)
                .append("count", 0)
                .append("_id", "test");
```
자세한 내용은 아래 각각의 reply 내용을 확인해주시면 될 것 같습니다.

> 혹시 제가 잘못 생각한 부분이 있을 수도 있습니다. 그런 부분은 절 태그해서 이야기해주시면 답변 드리겠습니다.

BlaCkinkGJ commented on src/test/java/com/team04/musiccloud/db/AccountCustomRepositoryTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test
+    public void updateAccount() {
+        accountCustomRepository.updateAccount("test@test.com", "test", "test", false);
```
codacy에서 대부분의 항목에 대해서 아래와 같은 문제를 일으키고 있습니다.

![image](https://user-images.githubusercontent.com/16631264/59005826-f24b5180-8859-11e9-8149-e2f3e95b4eb9.png)

위 내용에서도 알겠지만 update만 수행하기 때문에 test를 어떻게 할 지 난처할 가능성이 높습니다. 이 경우 제가 추천하는 방법은 아래와 같습니다.

```java
try {
  account ~~~
} catch(Exception e) {
  fail();
}
```
즉, 예외 처리가 나면 fail로 처리하도록 하는 것입니다. 이렇게 하면 대다수 현재 codacy에서 발생하는 문제를 해결 가능할 것이라 사료됩니다.

BlaCkinkGJ reviewed (changes requested):
Codacy 관련 부분에 대한 수정 조언입니다.

deepseafishy committed:
Merge branch 'master' of https://github.com/pnu-004-team4/team04 into DaOnPark

deepseafishy committed:
Fixed JUnit test cases for codacy issues

Implemented JUnit test cases for code coverage. Haven't covered all the parts, but all of them are covered at least 84%. Some test cases were fixed due to codacy issues. Couple of duplicate codes were also handled.

AudioMetaDao.java & FileMetaDao.java
1. Reverted to the deleteDataInDatabase() method for handling duplication.

MetadataCustomRepository.java
1. deleteMetadata() method now returns boolean.

deepseafishy committed:
Fixed JUnit test cases for Travis CI build issue

AudioMetaConverterTest had issues when built.

AudioMetaConverterTest.java
1. AudioMetaConverterTest's toDocument() method testing was wrong due to invalid ObjectID value. Appropriate hex string value was given to fix this issue.
